UnleashClient/strategies/Strategies.py

```diff
@@ -0,0 +1,116 @@
+# pylint: disable=dangerous-default-value
+from UnleashClient.constraints import Constraint
+
+
+class Strategy():
+    """
+    OG Strategy object.  This is the parent class for default and custom strategies.
+
+    In general, default & custom classes should only need to override:
+    * __call__() - Implementation of the strategy.
+    * load_provisioning - Loads strategy provisioning
+    """
+    def __init__(self,
+                 parameters: dict = {}) -> None:
+        """
+        A generic strategy objects.
+
+        :param parameters: 'parameters' key from strategy section (...from feature section) of
+        /api/clients/features response
+        """
+        # Experiment information
+        self.parameters = parameters
+
+        self.parsed_provisioning = self.load_provisioning()
+
+    # pylint: disable=no-self-use
+    def load_provisioning(self) -> list:
+        """
+        Method to load data on object initialization, if desired.
+
+        This should parse the raw values in self.parameters into format Python can comprehend.
+        """
+        return []
+
+    def __eq__(self, other):
+        return self.parameters == other.parameters
+
+    def __call__(self, context: dict = None) -> bool:
+        """
+        Strategy implementation goes here.
+
+        :param context: Context information
+        :return:
+        """
+        return False
+
+
+class StrategyV2:
```

Comment:
I'm not sure if I understand the need to for StrategyV2 and not just adding this to the existing implementation?

Reply:
I waffled a bit back and forth.  I'm happy to revisit that decision (esp. since I made it around 3:30 AM).

My reasoning was that:
- Right now, checking a feature toggle aka running strategies invokes the `__call__` class method on each strategy.
- I wanted constraint checks to be part of [whatever runs when we check a strategy].
- The instructions for custom strategies goes to the effect: extend the Strategy class and override the `load_parameters` method (to parse server provisioning) and `__call__` class method (to apply the strategy).
- If possible, don't break old custom strategies.

(In retrospect, overriding a class method might not have been the best idea.)

I think there's three choices:
1. Keep the old `Strategy` object around and implement constraints in the `StrategyV2` object.  The strategy implementation moves to the `apply_strategy` method and the `__call__` method checks constraints -> applies strategies.  
2. Move `StrategyV2` changes to `Strategy` object.  Breaks existing CustomStrategies.
3. Do ^ but re-work existing implementation so [checking a strategy] uses a method other than `__call__` which continues to implements the strategy.  Supports backwards compatibility but leaves the API/architecture a little messy. :(

Thoughts/suggestions?

Reply:
Here are my 2 cents:

Given that this is a major release (with the freedom for breaking backwards compatibility that comes with that),  I'm tempted to suggest to go with option `2`. (Though if possible, break existing CustomStrategies with a descriptive error message on how to upgrade).

If that's not a desired option, I think option `1` would be a second best option.